app/controllers/coffee_shop_controller.rb

```diff
@@ -0,0 +1,79 @@
+# frozen_string_literal: true
+
+require 'sinatra/base'
+require 'json'
+require_relative '../services/coffee_shop_finder_service'
+
+class CoffeeShopController < Sinatra::Base
+  configure do
+    set :protection, except: [:host_authorization]
+    set :show_exceptions, false
+    set :raise_errors, true
+    enable :logging
+  end
+
+  before do
+    content_type :json
+  end
+
+  get '/closest_shops' do
+    content_type 'text/plain'
+    lat, lon = validate_coordinates!(params)
+
+    shops_with_distances = coffee_shop_finder_service.closest_shops(lat, lon)
+    format_response(shops_with_distances, lat, lon)
+  rescue StandardError => e
+    handle_error(e)
+  end
+
+  private
+
+  # Format shops into "Name --> distance <-- (user-lat, user_lon)" strings
+  def format_response(shops_with_distances, user_lat, user_lon)
+    header = "Coffee shops nearest (#{user_lat}, #{user_lon}) by distance:\n\n"
+
+    header + shops_with_distances.map do |shops_with_distance|
+      shop = shops_with_distance[:shop]
+      distance = shops_with_distance[:distance]
+
+      "#{distance} <--> #{shop.name}"
+    end.join("\n")
+  end
+
+  def validate_coordinates!(parameters)
+    error!(400, 'Invalid coordinates') unless parameters[:lat] && parameters[:lon]
+    error!(400, 'Coordinates must be numeric') unless numeric?(parameters[:lat]) && numeric?(parameters[:lon])
+
+    lat = parameters[:lat].to_f
+    lon = parameters[:lon].to_f
+    error!(400, 'Invalid coordinates') unless lat.between?(-90, 90) && lon.between?(-180, 180)
+
+    [lat, lon]
+  end
+
+  def numeric?(str)
+    return false unless str =~ /\A[-+]?[0-9]*\.?[0-9]+\Z/
+
+    Float(str)
+  end
+
+  # Handle errors with appropriate HTTP status codes
+  def handle_error(error)
+    status_code = case error.message
+                  when /Invalid CSV/ then 400
```

Comment:
400 is not correct here according to industry standards: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/400.

"The HTTP 400 Bad Request [client error response](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#client_error_responses) status code indicates that the server would not process the request due to something the server considered to be a client error. The reason for a 400 response is typically due to malformed request syntax, invalid request message framing, or deceptive request routing.

Clients that receive a 400 response should expect that repeating the request without modification will fail with the same error."

It's not requester fault that the server's data is invalid.

Reply:
The requester may not modify the request and stop receiving 400 at later times

Reply:
Yea correct this is a mistake. The fact that our DB data is invalid is not on the client side.
I should've use code `500` or [422 Unprocessable Content](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/422)  - "status code indicates that the server understood the content type of the request content, and the syntax of the request content was correct, but it was unable to process the contained instructions."